topdown/topdown.go

```diff
@@ -1823,14 +1831,30 @@ func evalRefRuleResultRecSet(t *Topdown, set *ast.Set, ref, path ast.Ref, iter f
 		for _, e := range *set {
 			undo := t.Bind(k, e.Value, nil)
 			path = append(path, e)
+
 			if err := evalRefRuleResultRec(t, e.Value, tail, path, iter); err != nil {
 				return err
 			}
 			path = path[:len(path)-1]
 			t.Unbind(undo)
 		}
 		return nil
+	case ast.Array, ast.Object, *ast.Set:
```

Comment:
This entire function can be simplified to unify the set elements with the next element in the ref. Something like...

```
	head, tail := ref[0], ref[1:]
	k := head.Value
	for _, e := range *set {
		undo, err := evalEqUnify(t, e.Value, k, nil, func(t *Topdown) error {
			return evalRefRuleResultRec(t, PlugValue(k, t.Binding), tail, path.Append(e), iter)
		})
		if err != nil {
			return err
		}
		t.Unbind(undo)
	}
	return nil
```